webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,99 @@
-type ExperimentsContext = {
-  experimentsCache: Map<string, ExperimentPagination>;
-  updateExperimentsCache: (ws: Map<string, ExperimentPagination>) => void;
-};
-
-const ExperimentsContext = createContext<ExperimentsContext | null>(null);
-
-export const ExperimentsProvider: React.FC<PropsWithChildren> = ({ children }) => {
-  const [experimentsCache, updateExperimentsCache] = useState<Map<string, ExperimentPagination>>(
-    Map<string, ExperimentPagination>(),
-  );
-  return (
-    <ExperimentsContext.Provider value={{ experimentsCache, updateExperimentsCache }}>
-      {children}
-    </ExperimentsContext.Provider>
-  );
-};
+// this is singleton
+export class ExperimentsService {
+  static #isInternalConstructing = false;
+  static #instance: ExperimentsService;
+  #experimentsCache: WritableObservable<Map<string, ExperimentsCache>> = observable(Map());
+  #experimentMap: WritableObservable<Map<number, ExperimentItem>> = observable(Map());
 
-export const useFetchExperiments = (
-  params: GetExperimentsParams,
-  canceler: AbortController,
-): (() => Promise<void>) => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  private constructor() {
+    if (!ExperimentsService.#isInternalConstructing) {
+      throw new TypeError('ExperimentsService is not constructable');
+    }
   }
-  const { experimentsCache, updateExperimentsCache } = context;
-
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getExperiments(params, { signal: canceler.signal });
-      updateExperimentsCache(experimentsCache.set(encodeParams(params), response));
-    } catch (e) {
-      handleError(e);
+
+  public static getInstance(): ExperimentsService {
+    if (!ExperimentsService.#instance) {
+      ExperimentsService.#isInternalConstructing = true;
+      ExperimentsService.#instance = new ExperimentsService();
+      ExperimentsService.#isInternalConstructing = false;
     }
-  }, [canceler, experimentsCache, params, updateExperimentsCache]);
-};
+    return ExperimentsService.#instance;
+  }
 
-export const useExperiments = (params: GetExperimentsParams): Loadable<ExperimentPagination> => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useExperiments outside of Experiment Context');
+  public experimentsByParams(
+    params: Readonly<GetExperimentsParams>,
+  ): Observable<Loadable<ExperimentPagination>> {
+    return this.#experimentsCache.select((map) => {
+      const cache = map.get(encodeParams(params));
+      if (!cache) {
+        return NotLoaded;
+      }
+      const experiments: ExperimentItem[] = [];
+      const expMap = this.#experimentMap.get();
+      for (const id of cache.experimentIds) {
+        const expItem = expMap.get(id);
+        if (expItem) {
+          experiments.push(expItem);
+        }
+      }
```

Comment:
nit: I think this could be replaced with 
`const experiments: ExperimentItem[] = cache.experimentIds.map(id => expMap.get(id)).filter(exp => !!exp)`.
It's slower, since there are two loops instead of one, but I think the functional approach is easier to parse. Up to you.

Reply:
<img width="815" alt="image" src="https://user-images.githubusercontent.com/109113616/215226733-24f2e674-82d2-46c3-a2c5-8273c746829d.png">

idk why ts complains, but i leave this due to the error

Reply:
You wanna use a flatMap here instead of .map.filter

Reply:
`flatMap` didnt work either. 

looks like its[ the reason ](https://stackoverflow.com/questions/64425065/removing-undefined-from-es6-mapping-without-typescript-errors)

in this case, using `as` should be safe, so added `as` at the end of `filter`

Reply:
`.flatMap(exp) => exp ? [exp] : [])`

Reply:
thats smart. updated